app/src/main/java/org/simple/clinic/home/overdue/removepatient/RemoveAppointmentSheetController.kt

```diff
@@ -0,0 +1,65 @@
+package org.simple.clinic.home.overdue.removepatient
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.ofType
+import io.reactivex.rxkotlin.withLatestFrom
+import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.overdue.AppointmentRepository
+import org.simple.clinic.widgets.UiEvent
+import javax.inject.Inject
+
+typealias Ui = RemoveAppointmentSheet
+typealias UiChange = (Ui) -> Unit
+
+class RemoveAppointmentSheetController @Inject constructor(
+    private val repository: AppointmentRepository
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(event: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = event.compose(ReportAnalyticsEvents()).replay().refCount()
+
+    return Observable.merge(
+        reasonClicks(replayedEvents),
+        doneClicks(replayedEvents))
+  }
+
+  private fun reasonClicks(events: Observable<UiEvent>): Observable<UiChange> {
+    val mergedReasons = Observable.merge(
+        events.ofType<AlreadyVisitedReasonClicked>(),
+        events.ofType<CancelReasonClicked>())
+
+    return mergedReasons
+        .distinctUntilChanged()
+        .map { { ui: Ui -> ui.enableDoneButton() } }
+  }
+
+  private fun doneClicks(events: Observable<UiEvent>): Observable<UiChange> {
+    val appointmentUuids = events.ofType<RemoveAppointmentSheetCreated>()
+        .map { it.appointmentUuid }
+
+    val mergedReasons = Observable.merge(
+        events.ofType<AlreadyVisitedReasonClicked>(),
+        events.ofType<CancelReasonClicked>())
+
+    val doneWithLatestFromReasons = events.ofType<RemoveReasonDoneClicked>()
+        .withLatestFrom(appointmentUuids, mergedReasons)
+
+    val markAsVisitedStream = doneWithLatestFromReasons
+        .filter { (_, _, reason) -> reason is AlreadyVisitedReasonClicked }
+        .flatMap { (_, uuid, _) ->
+          repository.markAppointmentAsVisited(uuid)
+              .andThen(Observable.just { ui: Ui -> ui.closeSheet() })
+        }
+
+    val cancelWithReasonStream = doneWithLatestFromReasons
+        .filter { (_, _, reason) -> reason is CancelReasonClicked }
+        .flatMap { (_, uuid, reason) ->
+          repository.cancelAppointmentWithReason(uuid, (reason as CancelReasonClicked).selectedReason)
+              .andThen(Observable.just { ui: Ui -> ui.closeSheet() })
+        }
+
+    return Observable.merge(markAsVisitedStream, cancelWithReasonStream)
```

Comment:
This Rx chain looks really nicely done. Is this the same chain you were talking about yesterday night? I wouldn't do it any differently.

Reply:
I was suspicious at first why are we first doing a merge if we're anyway filtering it later, but it makes sense. Removing merge and filter makes the tests fail so that's good.

Reply:
Yes. This is the chain. Relieved!